WalletWasabi/Stores/IndexStore.cs

```diff
@@ -23,8 +23,11 @@ namespace WalletWasabi.Stores
 	/// </summary>
 	public class IndexStore
 	{
-		public IndexStore(Network network, SmartHeaderChain hashChain)
+		public IndexStore(string workFolderPath, Network network, SmartHeaderChain hashChain)
 		{
+			WorkFolderPath = Guard.NotNullOrEmptyOrWhitespace(nameof(workFolderPath), workFolderPath, trim: true);
+			IoHelpers.EnsureDirectoryExists(WorkFolderPath);
```

Comment:
Same here.

Reply:
Why not put this into the InitializeAsync?

Reply:
This does not belong to `InitializeAsync`. Ideally, it should be done before the class is instantiated, then tests will be easier.

Reply:
Why not belongs there? Moving it out of the class makes no sense for me - the store should ensure their own things by itself like the folder.

Reply:
I find it more versatile when class does only as much is needed. Not everyone agrees on this one with me though - both approaches have some merits. For me personally:

* I don't like when constructors can throw exceptions. 
* I believe that parameters given to a class should be already valid.
* It makes mocking of the class harder as for a unit test you really need to touch filesystem and unit test should not need that.

Reply:
> I don't like when constructors can throw exceptions.

Can we say this: constructors would only throw when the creation of the instance is impossible? Guard and Nullable things can throw. On the other hand, constructors should not do any operations those are related with other objects and can fail - like creating directories. 

> It makes mocking of the class harder as for a unit test you really need to touch filesystem and unit test should not need that.

Agree.

Reply:
> Can we say this: constructors would only throw when the creation of the instance is impossible? Guard and Nullable things can throw. On the other hand, constructors should not do any operations those are related with other objects and can fail - like creating directories.

I think that guards are acceptable but with NRT, we should continue removing them because it allows the test code to create mocks. 

Do you have some specific example in mind to talk about something specific? I find it always helpful because general cases are hard to imagine.

Reply:
Regarding this case, let it be - as the same concept was used previously in BitcoinStore.

Reply:
No, I didn't have anything specific. Just that I also noticed that doing the initialization in the ctor is bad practice - so I was curious about your concepts.